design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+
+The Modularization rework should free this up a bit, but it's here that we'll want to actually try it out for Metal.  We'll see how far it can be pushed during this phase.  There may be bottlenecks in the non-rendering portion of the toolkit that assumed similar bottlenecks in the rendering that no longer exist.
+
+The goal here is to move as many setup stages as possible off the main thread.  That should include all the buffer allocation and setup and some of the indirect command buffer encoding.  Direct encoding is going to still occur on the main thread.
+
+#### Required Changes
+
+It's not clear what changes will need to be made here.  The process goes like this:
+- Find a good test case.  Usually something that involves a lot of zooming and panning.
+- Run the test case while recording one or more of: memory or CPU
+- Look for a bottleneck
+- Fix the bottleneck
+
+Typically bottlenecks are where the main CPU is waiting on something it shouldn't be, or tiles are only being loaded on a single thread or other similar issues.  It's never obvious until you start staring at the traces.
+
+#### Benefits
+
+MapLibre Native makes some use of threads, but not as much as it should.  With OpenGL it's not using shared contexts, though it does a lot of work in the Layers to prep for the GL work that has to be done on the main thread.  So it's not terrible, but it obviously could be better.
+
+Mobile devices, even the really cheap ones, now have a lot of CPU cores and a lot of opportunity for work to be done off the main thread.  Maps have an embarassing amount of high level parallelism.  It's worth periodically analyzing the performance to see what might be updated for data loading.
+
+The benefits are faster loading, lower latency, better responsiveness to the user.
+
+### Shader optimization
+
+The first version of shaders will work with the direct rendering in Metal.  They'll be simple manual translations from the GLSL in the existing toolkit and they'll look and act similarly to the old shaders.
+
+The second version of the shaders will consist of changes to support indirect rendering.  The way buffers are passed in is different when doing indirect rendering in a way that's not obvious.
+
+Further, this is the time to break the shaders to modular pieces to be reused by developers outside the toolkit.  Passing in a new shader to replace existing functionality works a lot better when you can call sub-routines to implement that MapLibre Native basics for you.
+
+#### Required Changes
+
+This pass is the last version of the shaders and would consist of:
+- Modularization.  We can make it easier to reuse shaders by sharing code where useful.  Even putting the core of a given shader into a subroutine would make it easier to reuse.
+- Data driven styling.  How much of that logic can be pushed into the shaders?  The more work they do, the less the CPU does.
```

Comment:
Calculation is very cheap for the shaders.  Recalculating the same value over and over again isn't a big problem.  If we do find some calculations that are expensive, they could be moved to calculation shaders and their results shared.  I thought we'd do that with WhirlyGlobe-Maply, but it never got to be a big enough cost.

It's more likely that memory fetching is the bigger problem.  That tends to be what takes up the time.

My intuition matched yours before we moved WhirlyGlobe-Maply to Metal.  GPUs just got a lot faster on these devices.  So now I believe we should do a lot more in the shaders.  We should move more of the data driven styling into the shaders, in particular.

As for modularization, I'd say the data driven styling should be presented to the rendering module in a coherent way, but how it implements that styling is up to the developer.

Reply:
The situation is radically different on iOS.  And that points to something I keep coming back to:
Let's specify how things are supposed to look and let the implementation decide how to do them.
- In WebGL, less calculation in the shader.  
- Maybe for Android, less calculation for low end devices.
- For iOS, more calculation in the shader.

Leave it up to the implementation of the SDK is what we're saying.

Reply:
> I was able to improve performance in our own renderers by making code a pre-processor conditional / generating shader variations. I also needed to do this (on desktop / laptop) in our maplibre-gl-js related code, because clients would complain about fan-noise. Especially on IGPs you can quickly become shader-bound.

I'm less of a fan at the idea of shader permutations. Especially if we go the precompiled shader route. Size of the library will rapidly grow out of control if we start creating dozens or hundreds of precompiled shader permutations. So we permute at runtime, but then that locks you out of precompiling. Lose Lose.

Reply:
I don't think there are enough shader variants to need this.  If/else blocks and switches aren't expensive as they once were.

Reply:
The real pain point is around those data-driven layers. We either replicate `#pragma mapbox` to allow for permutations of uniform<->vertex attribute or we greedily reserve the maximum allowable (minimum guaranteed) number of per-vertex attributes to interpret dynamically in the shader. For the bulk of drawables that is just going to waste VRAM (assuming Metal requires the vertex attribute layout specified in a bound shader reflect the vertex buffer being rasterized).

Reply:
Metal allows you to do it at least a couple of ways.  You can interpret things a bit more raw, just getting the buffers and doing what you like.  Or you can set up this fairly complex binding logic that resembles what OpenGL does... mostly.

For the latter case you can set up defaults.  So, for instance, color can be an array or it can be one value.  The shader doesn't know the difference... if you set it up that way.

Now part of this is my unfamiliarity with the low shader implementation in MapLibre Native.  What puzzles me is you can do this default value trick in OpenGL too.  So there must be some reason they didn't do that, perhaps the maximum number of available attributes.

Was this because of data driven styling?  If so, it might be an argument to move that logic into its own calculation shaders and then feed the results into much simpler render shaders.  Honestly, that's a decision to make during implementation.

Reply:
Coming back to this after thinking more, I kind of like the calculation shader idea.  It moves the cost of data driven styling into the pipeline itself and makes your render shaders much simpler.

It also provides a pathway for shaders to emit the kinds of data the render shaders expect.  So then you could imagine a shader which emits vectors to be widened with all the various inputs filled in by that calculation shader.

In concept this isn't dissimilar from putting together shaders on the fly.  Instead you're adding stages to the render pipeline on the fly.  Since all those pieces are pre-compiled and/or set up on another thread it's going to be pretty fast.  It also has the advantage of being scalable.  You could go nuts with the data driving styling past the point OpenGL could support.

Something to consider later in the implementation.  But it might be simpler than the pragma approach.